implementation of an AST -> SCFG transformer

esc opened the pull request:
AST: Python Abstract Synatx Tree
(S)CFG: (Structured) Control Flow Graph

Items to note for reviewers:

* Most of changes are in the `ast_transforms.py` module which contains the meat of the implementation.
* Temporary data structures are created for a pure AST based CFG (`ASTCFG`) with blocks that are writable. This made it much easier to implement the builder pattern. They can easily be converted to a SCFG using the `ASTCFG.to_SCFG()` function.
* The `AST2SCFGTranformer` is stateful and is populated with housekeeping data structures post transform.
* The tests are all in `test_ast_transforms.py`. They are of the usual format: Python function as input and serialized graph to compare to.
* The tests also check the values of the unreachable and empty nodes that have been pruned from the CFG.
* The for-loop decomposition is a bit involved, hence there is an extensive docstring to describe what it does.
* Some minor changes had to be made to the `scfg.py` module to support displaying "broken" SCFGs during development. This makes it easier to visually debug the graph upon making programming mistakes.
* The modules `basic_block.py` and `rendering.py` were modified to support construction and visualization of the `PythonASTBlock`, which is a new block type for the SCFG that contains a block derived from Python AST.
* The central entrypoint is: `AST2CFG(function) -> SCFG `. This will turn a Python function into an SCFG.
* A stub has been added for the invserse, SCFG -> AST transformer.

Example:

```python
from numba_rvsdg import AST2SCFG
def fun():
    ...
scfg = AST2SCFG(fun)
scfg.render()
scfg.restructure()
scfg.render()
```

esc commented:
@brandonwillard have a look at the above ☝️ may be of interest to you.

brandonwillard commented:
> @brandonwillard have a look at the above ☝️ may be of interest to you.

This is fantastic!

esc commented:
Coverage report for [203ec7f](https://github.com/numba/numba-rvsdg/pull/114/commits/203ec7f2585acf080dd334984dc9d09150314cbc)

```
coverage report
Name                                                Stmts   Miss  Cover
-----------------------------------------------------------------------
numba_rvsdg/__init__.py                                 1      0   100%
numba_rvsdg/core/datastructures/ast_transforms.py     197      8    96%
numba_rvsdg/core/datastructures/basic_block.py        106      2    98%
numba_rvsdg/core/datastructures/block_names.py         13      0   100%
numba_rvsdg/core/datastructures/byte_flow.py           17      0   100%
numba_rvsdg/core/datastructures/flow_info.py           46      0   100%
numba_rvsdg/core/datastructures/scfg.py               404     29    93%
numba_rvsdg/core/transformations.py                   309      3    99%
numba_rvsdg/core/utils.py                              24      1    96%
numba_rvsdg/networkx_vendored/scc.py                   64     27    58%
numba_rvsdg/rendering/rendering.py                    167    125    25%
numba_rvsdg/tests/mock_asm.py                         122      2    98%
numba_rvsdg/tests/simulator.py                        226     46    80%
numba_rvsdg/tests/test_ast_transforms.py              200    116    42%
numba_rvsdg/tests/test_byteflow.py                     68      3    96%
numba_rvsdg/tests/test_figures.py                      21      0   100%
numba_rvsdg/tests/test_mock_asm.py                     44      0   100%
numba_rvsdg/tests/test_scc.py                          47     38    19%
numba_rvsdg/tests/test_scfg.py                         44      5    89%
numba_rvsdg/tests/test_simulate.py                    114      2    98%
numba_rvsdg/tests/test_transforms.py                  171      1    99%
numba_rvsdg/tests/test_utils.py                        66      0   100%
-----------------------------------------------------------------------
TOTAL                                                2471    408    83%
```

sklam commented on numba_rvsdg/core/datastructures/ast_transforms.py:
```diff
@@ -0,0 +1,603 @@
+
+    def __init__(self, code: Callable[..., Any], prune: bool = True) -> None:
+        # Prune empty and unreachable blocks from the CFG.
+        self.prune: bool = prune
+        # Save the code for transformation.
+        self.code: Callable[..., Any] = code
+        # Monotonically increasing block index, 0 is reserved for genesis.
+        self.block_index: int = 1
+        # Dict mapping block indices as strings to WritableASTBlocks.
+        # (This is the data structure to hold the CFG.)
+        self.blocks: ASTCFG = ASTCFG()
+        # Initialize first (genesis) block, assume it's named zero.
+        # (This also initializes the self.current_block attribute.)
+        self.add_block(0)
+        # Stack for header and exiting block of current loop.
+        self.loop_stack: list[LoopIndices] = []
```
Type annotation for instance variables should be in class scope. See https://peps.python.org/pep-0526/#class-and-instance-variable-annotations

esc replied:
done in 8a119a4bc835362dc149a6bc4d85e28d972ec3cd

sklam commented on numba_rvsdg/core/datastructures/ast_transforms.py:
```diff
@@ -0,0 +1,603 @@
+    def transform(self) -> None:
+        """Transform Python function stored as self.code."""
+        # Convert source code into AST.
+        tree = ast.parse(textwrap.dedent(inspect.getsource(self.code))).body
```
The expected input should be an AST instead of code. Otherwise, if I start with a source string, I'll need to make a code object first. I'll suggest pull out this AST creation logic.

sklam replied:
For example, i tried to use my [AST fuzzer](https://github.com/sklam/jittery/blob/pythrankenstein/pythrankenstein.py) to test this but i ended up needing to change the code like:

```diff
@@ -221,7 +221,10 @@ class AST2SCFGTransformer:
     def transform(self) -> None:
         """Transform Python function stored as self.code."""
         # Convert source code into AST.
-        tree = ast.parse(textwrap.dedent(inspect.getsource(self.code))).body
+        if isinstance(self.code, str):
+            tree = ast.parse(self.code).body
+        else:
+            tree = ast.parse(textwrap.dedent(inspect.getsource(self.code))).body
         # Assert that the code handed in was a function, we can only transform
         # functions.
         assert isinstance(tree[0], ast.FunctionDef)
```

Btw, the fuzzer will need `ast.Pass` handled to work correctly.

esc replied:
`ast.Pass` can be handled and pruned with the `prune_noop` pass: dc64724bc05890004065a97161d8130a88d7135a

esc replied:
I opted to include this snippet such that you can now feed both a callable and a string to the transformer: 17525f17ff64f36dc80f31a78194ef6921bacc3c

sklam reviewed (commented):
Just some minor adjustments. The biggest thing is for loop `orelse` need to be sealed.

esc commented on numba_rvsdg/core/datastructures/ast_transforms.py:
```diff
@@ -0,0 +1,603 @@
+        self.codegen(ast.parse(else_code).body)
+
+        # Recurs into the body of the else-branch.
+        self.codegen(node.orelse)
+
```
91781acef59facb77caf45b32283a4c67f145a5f

This took some more work to hit.

esc commented:
@sklam as discussed OOB, I have implemented `while: else:` and simplified the `handle_while` method: 85a6526de240ede75fa63dd99bfeaced6f665e5f

esc commented:
@sklam I have addressed all the review comments and added most of the features

The only thing I don't know how to decide on is how to further structure the API. In any case, the transformations are implemented and work so maybe we need to defer decisions about the API to a later stage.

sklam reviewed (commented):
Some nitpicks regarding accept AST as input. Everything else looks good.

esc commented on numba_rvsdg/core/datastructures/ast_transforms.py:
```diff
@@ -175,22 +195,46 @@ class AST2SCFGTransformer:
+    def unparse_code(code: str | Callable[..., Any]) -> list[type[ast.AST]]:
+        # Convert source code into AST.
+        if isinstance(code, str):
+            tree = ast.parse(code).body
```
now it is: 4b7d46b384d145f096bb51e8ee8b7fdb3f8d33e5

esc commented on numba_rvsdg/core/datastructures/ast_transforms.py:
```diff
@@ -175,22 +195,46 @@ class AST2SCFGTransformer:
+    loop_stack: list[LoopIndices]
+
+    def __init__(
+        self, code: str | Callable[..., Any], prune: bool = True
```
how about `list[ast.FunctionDef]` it's what we end up asserting on.

esc replied:
done in 41cc5e3c117e319b4c54478638f6a6a18311c766
